File: CS/ResourceForm.cs
using System;
using System.Drawing;
using System.Windows.Forms;
using DevExpress.XtraScheduler;

namespace SchedulerResourcesManagement {
    public partial class ResourceForm : Form {
        private readonly Resource sourceResource = null;
        private readonly Resource editedResourceCopy = new Resource();

        protected virtual Resource SourceResource { get { return sourceResource; } }
        protected virtual Resource EditedResourceCopy { get { return editedResourceCopy; } }

        protected virtual Image DefaultResourceImage { get { return SchedulerResourcesManagement.Properties.Resources.NoImage; } }

        public ResourceForm(Resource resource) {
            if (resource == null)
                throw new ArgumentNullException("resource");

            this.Disposed += new EventHandler(ResourceForm_Disposed);

            this.sourceResource = resource;
            InitializeComponent();

            UpdateEditedResourceCopy();
            UpdateForm();
        }

        void ResourceForm_Disposed(object sender, EventArgs e) {
            if (EditedResourceCopy != null)
                EditedResourceCopy.Dispose();
        }

        protected virtual void UpdateEditedResourceCopy() {
            EditedResourceCopy.Caption = SourceResource.Caption;
            EditedResourceCopy.Color = SourceResource.Color;
            EditedResourceCopy.Image = SourceResource.Image;
        }

        protected virtual void UpdateForm() {
            tbCaption.Text = EditedResourceCopy.Caption;
            lblColor.BackColor = EditedResourceCopy.Color;
            pbImage.Image = (EditedResourceCopy.Image == null ? DefaultResourceImage : EditedResourceCopy.Image);
            UpdateFormTitle();
        }

        protected virtual void UpdateFormTitle() {
            string formatArgument = string.Empty;
           
            if (!string.IsNullOrEmpty(tbCaption.Text))
                formatArgument = tbCaption.Text + " - ";

            this.Text = string.Format("{0}Resource", formatArgument);
        }

        protected virtual void ApplyChanges() {
            SourceResource.Caption = EditedResourceCopy.Caption;
            SourceResource.Color = EditedResourceCopy.Color;
            SourceResource.Image = EditedResourceCopy.Image;
        }

        private void btnColor_Click(object sender, EventArgs e) {
            if (colorDialog1.ShowDialog() == DialogResult.OK) {
                lblColor.BackColor = colorDialog1.Color;
                EditedResourceCopy.Color = colorDialog1.Color;
            }
        }

        private void pbImage_Click(object sender, EventArgs e) {
            if (openFileDialog1.ShowDialog() == DialogResult.OK) {
                pbImage.Image = Image.FromFile(openFileDialog1.FileName);
                EditedResourceCopy.Image = pbImage.Image;
            }
        }

        private void tbCaption_Validated(object sender, EventArgs e) {
            EditedResourceCopy.Caption = tbCaption.Text;
        }

        private void tbCaption_TextChanged(object sender, EventArgs e) {
            UpdateFormTitle();
        }

        private void btnOk_Click(object sender, EventArgs e) {
            ApplyChanges();
        }

    }
}

File: VB/ResourceForm.vb
Imports System
Imports System.Drawing
Imports System.Windows.Forms
Imports DevExpress.XtraScheduler

Namespace SchedulerResourcesManagement

    Public Partial Class ResourceForm
        Inherits Form

        Private ReadOnly sourceResourceField As Resource = Nothing

        Private ReadOnly editedResourceCopyField As Resource = New Resource()

        Protected Overridable ReadOnly Property SourceResource As Resource
            Get
                Return sourceResourceField
            End Get
        End Property

        Protected Overridable ReadOnly Property EditedResourceCopy As Resource
            Get
                Return editedResourceCopyField
            End Get
        End Property

        Protected Overridable ReadOnly Property DefaultResourceImage As Image
            Get
                Return Properties.Resources.NoImage
            End Get
        End Property

        Public Sub New(ByVal resource As Resource)
            If resource Is Nothing Then Throw New ArgumentNullException("resource")
            AddHandler Disposed, New EventHandler(AddressOf ResourceForm_Disposed)
            sourceResourceField = resource
            InitializeComponent()
            UpdateEditedResourceCopy()
            UpdateForm()
        End Sub

        Private Sub ResourceForm_Disposed(ByVal sender As Object, ByVal e As EventArgs)
            If EditedResourceCopy IsNot Nothing Then EditedResourceCopy.Dispose()
        End Sub

        Protected Overridable Sub UpdateEditedResourceCopy()
            EditedResourceCopy.Caption = SourceResource.Caption
            EditedResourceCopy.Color = SourceResource.Color
            EditedResourceCopy.Image = SourceResource.Image
        End Sub

        Protected Overridable Sub UpdateForm()
            tbCaption.Text = EditedResourceCopy.Caption
            lblColor.BackColor = EditedResourceCopy.Color
            pbImage.Image = If(EditedResourceCopy.Image Is Nothing, DefaultResourceImage, EditedResourceCopy.Image)
            UpdateFormTitle()
        End Sub

        Protected Overridable Sub UpdateFormTitle()
            Dim formatArgument As String = String.Empty
            If Not String.IsNullOrEmpty(tbCaption.Text) Then formatArgument = tbCaption.Text & " - "
            Text = String.Format("{0}Resource", formatArgument)
        End Sub

        Protected Overridable Sub ApplyChanges()
            SourceResource.Caption = EditedResourceCopy.Caption
            SourceResource.Color = EditedResourceCopy.Color
            SourceResource.Image = EditedResourceCopy.Image
        End Sub

        Private Sub btnColor_Click(ByVal sender As Object, ByVal e As EventArgs)
            If colorDialog1.ShowDialog() = DialogResult.OK Then
                lblColor.BackColor = colorDialog1.Color
                EditedResourceCopy.Color = colorDialog1.Color
            End If
        End Sub

        Private Sub pbImage_Click(ByVal sender As Object, ByVal e As EventArgs)
            If openFileDialog1.ShowDialog() = DialogResult.OK Then
                pbImage.Image = Image.FromFile(openFileDialog1.FileName)
                EditedResourceCopy.Image = pbImage.Image
            End If
        End Sub

        Private Sub tbCaption_Validated(ByVal sender As Object, ByVal e As EventArgs)
            EditedResourceCopy.Caption = tbCaption.Text
        End Sub

        Private Sub tbCaption_TextChanged(ByVal sender As Object, ByVal e As EventArgs)
            UpdateFormTitle()
        End Sub

        Private Sub btnOk_Click(ByVal sender As Object, ByVal e As EventArgs)
            ApplyChanges()
        End Sub
    End Class
End Namespace
